scripts/binary_to_array.py

```diff
@@ -0,0 +1,289 @@
+        "namespace %s {" % ns for ns in namespaces
+    ])
+  else:
+    data.extend([
+        "#if defined(__cplusplus)",
+        "extern \"C\" {",
+        "#endif  // defined(__cplusplus)"])
+
+  data.extend([
+      "",
+      "extern const size_t %s;" % array_size_name,
+      "extern const char %s[];" % fileid,
+      "extern const unsigned char %s[];" % array_name, "",
+      "const unsigned char %s[] = {" % array_name
+  ])
+  length = len(input_bytes)
+  line = ""
+  for idx in range(0, length):
+    if idx % WIDTH == 0:
+      line += "  "
+    else:
+      line += " "
+    line += "0x%02x," % input_bytes[idx]
+    if idx % WIDTH == WIDTH - 1:
+      data.append(line)
+      line = ""
+  data.append(line)
+  data.append("  0x00  // Extra \\0 to make it a C string")
+
+  data.extend([
+      "};",
+      "",
+      "const size_t %s =" % array_size_name,
+      "  sizeof(%s) - 1;" % array_name,
+      "",
+      "const char %s[] =" % fileid,
+      "  \"%s\";" % filename,
+      "",
+  ])
+
+  if namespaces:
+    data.extend([
+        "}  // namespace %s" % ns for ns in namespaces
+    ][::-1])  # close namespaces in reverse order
+  else:
+    data.extend([
+        "#if defined(__cplusplus)",
+        "}  // extern \"C\"",
+        "#endif  // defined(__cplusplus)"
+    ])
+  data.extend([
+      ""
+  ])
+  return data
+
+
+def main():
+  """Read an binary input file and output to a C/C++ source file as an array.
+
+  Raises:
+    Error: If an input file is not specified.
+  """
+
+  args = arg_parser.parse_args()
+
+  input_file = args.input
+  if not input_file:
+    raise Error("No input file specified.")
+  input_file_base = args.input.rsplit(".", 1)[0]
+
+  output_source = args.output_source
+  if not output_source:
+    output_source = input_file_base + ".cc"
+    logging.debug("Using default --output_source='%s'", output_source)
+
+  output_header = args.output_header
+  if not output_header:
+    output_header = input_file_base + ".h"
+    logging.debug("Using default --output_header='%s'", output_header)
+
+  identifier_base = sub("[^0-9a-zA-Z]+", "_", path.basename(input_file_base))
+  array_name = args.array
+  if not array_name:
+    array_name = identifier_base + "_data"
+    logging.debug("Using default --array='%s'", array_name)
+
+  array_size_name = args.array_size
+  if not array_size_name:
+    array_size_name = identifier_base + "_size"
+    logging.debug("Using default --array_size='%s'", array_size_name)
+
+  fileid = args.filename_identifier
+  if not fileid:
+    fileid = identifier_base + "_filename"
+    logging.debug("Using default --filename_identifier='%s'", fileid)
+
+  filename = args.filename
+  if filename is None:  # but not if it's the empty string
+    filename = path.basename(input_file)
+    logging.debug("Using default --filename='%s'", filename)
+
+  header_guard = args.header_guard
+  if not header_guard:
+    header_guard = sub("[^0-9a-zA-Z]+", "_", "_" + output_header).upper()
```

Comment:
I suppose this could end up generating double underscores, which are reserved.

If you care, you could add another sub() call to fix that.

Optional.

Reply:
Done. Also removed the leading underscore which made the identifier non-conforming unconditionally.